Possible flaky tests.

Recording these in case they reoccur.

First failure:
```
======================================================================
ERROR: test_get_new_journals (tests.test_flashflood.TestFlashFlood)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/xbrianh/flash-flood/tests/test_flashflood.py", line 188, in test_get_new_journals
    new_journals = [c for c in self.flashflood._new_journals()]
  File "/home/travis/build/xbrianh/flash-flood/tests/test_flashflood.py", line 188, in <listcomp>
    new_journals = [c for c in self.flashflood._new_journals()]
  File "/home/travis/build/xbrianh/flash-flood/flashflood/__init__.py", line 254, in _new_journals
    yield self._get_journal(_JournalID.from_key(item.key))
  File "/home/travis/build/xbrianh/flash-flood/flashflood/__init__.py", line 250, in _get_journal
    return _Journal(journal_id, self._get_manifest(journal_id), body)
  File "/home/travis/build/xbrianh/flash-flood/flashflood/__init__.py", line 245, in _get_manifest
    return json.loads(self.bucket.Object(key).get()['Body'].read().decode("utf-8"))
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/boto3/resources/factory.py", line 520, in do_action
    response = action(self, *args, **kwargs)
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/boto3/resources/action.py", line 83, in __call__
    response = getattr(parent.meta.client, operation_name)(**params)
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/botocore/client.py", line 357, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/botocore/client.py", line 661, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.errorfactory.NoSuchKey: An error occurred (NoSuchKey) when calling the GetObject operation: The specified key does not exist.
```

After restarting the test a second failure:
```
======================================================================
FAIL: test_get_event (tests.test_flashflood.TestFlashFlood) [Get event after journaling]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/xbrianh/flash-flood/tests/test_flashflood.py", line 49, in test_get_event
    self.assertTrue(self.flashflood.event_exists(event_id))
AssertionError: False is not true
```

During this PR: https://github.com/xbrianh/flash-flood/pull/24
This travis build (passed on the third run): https://travis-ci.org/xbrianh/flash-flood/builds/595241129?utm_source=github_status&utm_medium=notification

--- a/Makefile
+++ b/Makefile
@@ -1,10 +1,13 @@
-.PHONY: test lint tests clean build install
+.PHONY: test lint mypy tests clean build install
 MODULES=flashflood tests
 
-test: lint tests
+test: lint mypy tests
 
 lint:
 	flake8 $(MODULES) *.py
+
+mypy:
+	mypy --ignore-missing-imports $(MODULES)
 
 tests:
 	PYTHONWARNINGS=ignore:ResourceWarning coverage run --source=flashflood \

--- a/flashflood/__init__.py
+++ b/flashflood/__init__.py
@@ -1,51 +1,37 @@
 import io
 from datetime import datetime
 import json
+import typing
 import requests
 from uuid import uuid4
 from concurrent.futures import ThreadPoolExecutor, as_completed
-from collections import namedtuple
 
 from flashflood.util import datetime_to_timestamp, datetime_from_timestamp, DateRange, S3Deleter
+from flashflood.objects import BaseJournal, BaseJournalUpdate
+from flashflood.identifiers import JournalID, TOMBSTONE_SUFFIX
+from flashflood.event import Event
+from flashflood.key_index import BaseKeyIndex
+from flashflood.exceptions import FlashFloodEventNotFound, FlashFloodJournalingError
 
 
-Event = namedtuple("Event", "event_id date data")
-_Journal = namedtuple("_Journal", "id_ manifest body")
+# TODO:
+# What happens if event updates are written for a journal that has been removed?
+#
+# It's possible to resolve journals for update markers that are out-of-date using the event index
+#
+# Event updates should be applied idempotently
+#
+# write tests for multiple updates/deletes to the same event
+# Create resolution policy for multiple updates/deletes to the same event
+#
+# Use bucket lifecycle policy for garbage collection
+# Warn if policy is not set upon FF instantiation
+#
+# Include size/len(events) metadat in journal listing?
 
-class _JournalID(str):
-    DELIMITER = "--"
-
-    @classmethod
-    def make(cls, start_timestamp, end_timestamp, blob_id):
-        end_timestamp = end_timestamp or "new"
-        return cls(start_timestamp + cls.DELIMITER + end_timestamp + cls.DELIMITER + blob_id)
-
-    @classmethod
-    def from_key(cls, key):
-        return cls(key.rsplit("/", 1)[1])
-
-    def _parts(self):
-        start_timestamp, end_timestamp, blob_id = self.split(self.DELIMITER)
-        return start_timestamp, end_timestamp, blob_id
-
-    @property
-    def blob_id(self):
-        return self._parts()[2]
-
-    @property
-    def start_date(self):
-        return datetime_from_timestamp(self._parts()[0])
-
-    @property
-    def end_date(self):
-        end_date = self._parts()[1]
-        if "new" == end_date:
-            return self.start_date
-        else:
-            return datetime_from_timestamp(end_date)
 
 class FlashFlood:
-    def __init__(self, s3_resource, bucket, root_prefix):
+    def __init__(self, s3_resource: typing.Any, bucket: str, root_prefix: str):
         self.s3 = s3_resource
         self.bucket = self.s3.Bucket(bucket)
         if root_prefix.endswith("/"):
@@ -53,257 +39,188 @@ class FlashFlood:
         self.root_prefix = root_prefix
         self._journal_pfx = f"{root_prefix}/journals"
         self._blobs_pfx = f"{root_prefix}/blobs"
-        self._new_pfx = f"{root_prefix}/new"
         self._update_pfx = f"{root_prefix}/update"
         self._index_pfx = f"{root_prefix}/index"
 
-    def put(self, data, event_id: str=None, date: datetime=None):
+        class _Journal(BaseJournal):
+            bucket = self.bucket
+            _journal_pfx = self._journal_pfx
+            _blobs_pfx = self._blobs_pfx
+
+        class _JournalUpdate(BaseJournalUpdate):
+            bucket = self.bucket
+            _pfx = self._update_pfx
+
+        class _KeyIndex(BaseKeyIndex):
+            bucket = self.bucket
+            _pfx = self._index_pfx
+
+        self._Journal = _Journal
+        self._JournalUpdate = _JournalUpdate
+        self._KeyIndex = _KeyIndex
+
+    def put(self, data, event_id: str=None, date: datetime=None) -> Event:
         """
         Write new event. If event exists, write update marker for event.
         """
         date = date or datetime.utcnow()
         timestamp = datetime_to_timestamp(date)
         event_id = event_id or str(uuid4())
-        assert _JournalID.DELIMITER not in event_id
+        assert JournalID.DELIMITER not in event_id
         if self.event_exists(event_id):
-            self._upload_update_action(event_id, "update", data)
+            journal_id = self._journal_for_event(event_id)
+            self._JournalUpdate.upload_update(journal_id, event_id, data)
             return Event(event_id, None, data)
         else:
-            blob_id = str(uuid4())
-            journal_id = _JournalID.make(timestamp, None, blob_id)
-            manifest = dict(journal_id=journal_id,
-                            from_date=timestamp,
-                            to_date=timestamp,
-                            size=len(data),
-                            events=[dict(event_id=event_id, timestamp=timestamp, offset=0, size=len(data))])
-            self._upload_journal(_Journal(journal_id, manifest, io.BytesIO(data)), is_new=True)
+            events = [dict(event_id=event_id, timestamp=timestamp, offset=0, size=len(data))]
+            journal = self._Journal(events, data=data, version="new")
+            journal.upload()
+            self._index_journal(journal)
+            print("new journal", journal.id_)
             return Event(event_id, date, data)
 
-    def delete(self, event_id):
+    def _index_journal(self, journal: BaseJournal):
+        journal_id = journal.id_
+        self._KeyIndex.put_batch({e['event_id']: journal_id for e in journal.events})
+
+    def delete(self, event_id: str):
         """
         Write delete marker for event.
         """
-        self._journal_for_event(event_id)  # raises FlashFloodEventNotFound if event not found
-        self._upload_update_action(event_id, "delete")
+        journal_id = self._journal_for_event(event_id)
+        self._JournalUpdate.upload_delete(journal_id, event_id)
+        # Deindexing immediately means the event is not available for lookup, but it will still appear in replay.
+        self._KeyIndex.delete(event_id)
 
-    def _upload_update_action(self, event_id, action, new_event_data=b""):
-        key = f"{self._update_pfx}/{event_id}{_JournalID.DELIMITER}{action}"
-        self.bucket.Object(key).upload_fileobj(io.BytesIO(new_event_data))
-
-    def update(self):
+    def update(self, number_of_updates_to_apply: int=1000) -> int:
         """
         Apply updates for each update and delete marker.
         """
-        for item in self.bucket.objects.filter(Prefix=self._update_pfx):
-            update_id = item.key.rsplit("/", 1)[1]
-            event_id, action = update_id.rsplit(_JournalID.DELIMITER, 1)
-            try:
-                journal_id = self._journal_for_event(event_id)
-            except FlashFloodEventNotFound:
-                continue
-            self._update_journal(journal_id, event_id, action, item.key)
-            print(f"Updated journal {journal_id}")
+        count = 0
+        for journal_id, updates in self._JournalUpdate.get_updates_for_all_journals():
+            print("updating journal", journal_id)
+            journal = self._Journal.from_id(journal_id)
+            new_journal = journal.updated(updates)
+            if new_journal != journal:
+                if not new_journal.is_empty:
+                    new_journal.upload()
+                    self._index_journal(new_journal)
+                journal.delete()
+            for u in updates.values():
+                u.delete()
+            count += len(updates)
+            if number_of_updates_to_apply <= count:
+                break
+        return count
 
-    def _update_journal(self, journal_id, event_id, action, update_marker_key):
-        with S3Deleter(self.bucket) as s3d:
-            journal = self._get_journal(journal_id)
-            journal_data = b""
-            events = list()
-            for e in journal.manifest['events']:
-                event_data = journal.body.read(e['size'])
-                e['offset'] = len(journal_data)
-                if event_id != e['event_id']:
-                    journal_data += event_data
-                    events.append(e)
-                elif "update" == action:
-                    event_data = self.bucket.Object(update_marker_key).get()['Body'].read()
-                    journal_data += event_data
-                    e['size'] = len(event_data)
-                    events.append(e)
-                elif "delete" == action:
-                    s3d.delete(f"{self._index_pfx}/{e['event_id']}")
-                else:
-                    raise Exception(f"No handler for journal update '{action}'")
-            if not journal_data:
-                self._delete_journals([journal.id_])
-            else:
-                from_timestamp = events[0]['timestamp']
-                to_timestamp = events[-1]['timestamp']
-                if from_timestamp == to_timestamp:
-                    to_timestamp = "new"
-                journal.manifest['events'] = events
-                journal.manifest['size'] = len(journal_data)
-                journal.manifest['from_date'] = from_timestamp
-                journal.manifest['to_date'] = to_timestamp
-                new_journal_id = _JournalID.make(from_timestamp, to_timestamp, journal.id_.blob_id)
-                self._upload_journal(_Journal(new_journal_id, journal.manifest, io.BytesIO(journal_data)))
-                if new_journal_id != journal_id:
-                    s3d.delete(f"{self._journal_pfx}/{journal_id}")
-
-    def journal(self, minimum_number_of_events=100, minimum_size=None):
-        assert 2 <= minimum_number_of_events
+    def journal(self, minimum_number_of_events: int=100, minimum_size: int=None):
         minimum_size = minimum_size or 0
-        journals_to_combine = list()
         number_of_events, size = 0, 0
-        for journal in self._new_journals():
-            size += journal.manifest['size']
-            number_of_events += len(journal.manifest['events'])
+        journals_to_combine = list()
+        for journal_id in self._new_journals():
+            journal = self._Journal.from_id(journal_id)
+            size += journal.size
+            number_of_events += len(journal.events)
             journals_to_combine.append(journal)
+            print("Found journal to combine", journal.id_)
             if minimum_number_of_events <= number_of_events and minimum_size <= size:
                 break
         if minimum_number_of_events > number_of_events:
             raise FlashFloodJournalingError(f"Journal condition: minimum_number_of_events={minimum_number_of_events}")
         if minimum_size > size:
             raise FlashFloodJournalingError(f"Journal condition: minimum_size={minimum_size}")
-        manifest = self.combine_journals(journals_to_combine)
-        print("Created new journal", manifest['journal_id'])
+        self.combine_journals(journals_to_combine)
 
-    def combine_journals(self, journals_to_combine):
-        events = list()
-        combined_data = b""
+    def _new_journals(self) -> typing.Iterator[JournalID]:
+        for journal_id in self._Journal.list():
+            if "new" == journal_id.version:
+                yield journal_id
+
+    def combine_journals(self, journals_to_combine: typing.List[BaseJournal]) -> BaseJournal:
+        new_journal = self._Journal()
+        objects_to_delete = journals_to_combine.copy()
         for journal in journals_to_combine:
-            for e in journal.manifest['events']:
-                events.append({**e, **dict(offset=len(combined_data))})
-            combined_data += journal.body.read()
-        blob_id = str(uuid4())
-        journal_id = _JournalID.make(events[0]['timestamp'], events[-1]['timestamp'], blob_id)
-        manifest = dict(journal_id=journal_id,
-                        from_date=events[0]['timestamp'],
-                        to_date=events[-1]['timestamp'],
-                        size=len(combined_data),
-                        events=events)
-        self._upload_journal(_Journal(journal_id, manifest, io.BytesIO(combined_data)))
-        self._delete_journals([journal.id_ for journal in journals_to_combine])
-        return manifest
+            print("combining journal", journal.id_)
+            updates = self._JournalUpdate.get_updates_for_journal(journal.id_)
+            objects_to_delete.extend(list(updates.values()))
+            journal = journal.updated(updates)
+            for e in journal.events:
+                new_offset = e['offset'] + len(new_journal.data)
+                new_journal.events.append({**e, **dict(offset=new_offset)})
+            new_journal.data += journal.body.read()
+        if not new_journal.is_empty:
+            new_journal.upload()
+            self._index_journal(new_journal)
+        for o in objects_to_delete:
+            o.delete()
+        return new_journal
 
-    def replay(self, from_date=None, to_date=None):
+    def replay(self, from_date: datetime=None, to_date: datetime=None) -> typing.Iterator[Event]:
         search_range = DateRange(from_date, to_date)
-        for journal_id in self._journal_ids(from_date, to_date):
-            journal = self._get_journal(journal_id)
-            for item in journal.manifest['events']:
+        for journal_id in self.list_journals(from_date, to_date):
+            print("replaying from journal", journal_id)
+            journal = self._Journal.from_id(journal_id)
+            for item in journal.events:
                 event_date = datetime_from_timestamp(item['timestamp'])
                 if event_date in search_range:
                     yield Event(item['event_id'], event_date, journal.body.read(item['size']))
                 elif event_date in search_range.future:
                     break
 
-    def _journal_for_event(self, event_id):
-        try:
-            key = next(iter(self.bucket.objects.filter(Prefix=f"{self._index_pfx}/{event_id}"))).key
-        except StopIteration:
-            raise FlashFloodEventNotFound()
-        return _JournalID(self.bucket.Object(key).metadata['journal_id'])
-
-    def _lookup_event(self, event_id):
-        journal_id = self._journal_for_event(event_id)
-        manifest = self._get_manifest(journal_id)
-        for item in manifest['events']:
-            if event_id == item['event_id']:
-                break
+    def _journal_for_event(self, event_id: str) -> JournalID:
+        journal_id = self._KeyIndex.get(event_id)
+        if journal_id is None:
+            raise FlashFloodEventNotFound(f"journal not found for {event_id}")
         else:
-            raise FlashFloodException(f"Event {event_id} not found in {journal_id}")
-        return journal_id, manifest, item
+            return journal_id
 
-    def event_exists(self, event_id):
-        try:
-            self._journal_for_event(event_id)
-            return True
-        except FlashFloodEventNotFound:
-            return False
+    def event_exists(self, event_id: str) -> bool:
+        return self._KeyIndex.get(event_id) is not None
 
-    def get_event(self, event_id):
-        journal_id, manifest, item = self._lookup_event(event_id)
-        blob_key = f"{self._blobs_pfx}/{journal_id.blob_id}"
-        byte_range = f"bytes={item['offset']}-{item['offset'] + item['size'] - 1}"
-        data = self.bucket.Object(blob_key).get(Range=byte_range)['Body'].read()
-        return Event(event_id, datetime_from_timestamp(item['timestamp']), data)
+    def get_event(self, event_id: str) -> Event:
+        journal_id = self._journal_for_event(event_id)
+        journal = self._Journal.from_id(journal_id)
+        return journal.get_event(event_id)
 
-    def replay_urls(self, from_date=None, to_date=None, maximum_number_of_results=1):
-        urls = list()
-        for journal_id in self._journal_ids(from_date, to_date):
-            manifest = self._get_manifest(journal_id)
-            journal_url = self._generate_presigned_url(journal_id)
-            urls.append(dict(manifest=manifest, events=journal_url))
-            if len(urls) == maximum_number_of_results:
-                break
-        return urls
-
-    def _upload_journal(self, journal, is_new=False):
-        key = f"{self._journal_pfx}/{journal.id_}"
-        blob_key = f"{self._blobs_pfx}/{journal.id_.blob_id}"
-        self.bucket.Object(blob_key).upload_fileobj(journal.body,
-                                                    ExtraArgs=dict(Metadata=dict(journal_id=journal.id_)))
-        self.bucket.Object(key).upload_fileobj(io.BytesIO(json.dumps(journal.manifest).encode("utf-8")))
-        if is_new:
-            self.bucket.Object(f"{self._new_pfx}/{journal.id_}").upload_fileobj(io.BytesIO(b""))
-        for item in journal.manifest['events']:
-            key = f"{self._index_pfx}/{item['event_id']}"
-            self.bucket.Object(key).upload_fileobj(io.BytesIO(b""),
-                                                   ExtraArgs=dict(Metadata=dict(journal_id=journal.id_)))
-
-    def _get_manifest(self, journal_id):
-        key = f"{self._journal_pfx}/{journal_id}"
-        return json.loads(self.bucket.Object(key).get()['Body'].read().decode("utf-8"))
-
-    def _get_journal(self, journal_id):
+    def _generate_presigned_url(self, journal_id: JournalID):
         key = f"{self._blobs_pfx}/{journal_id.blob_id}"
-        body = self.bucket.Object(key).get()['Body']
-        return _Journal(journal_id, self._get_manifest(journal_id), body)
+        client = self.s3.meta.client
+        return client.generate_presigned_url(ClientMethod="get_object",
+                                             Params=dict(Bucket=self.bucket.name, Key=key))
 
-    def _new_journals(self):
-        for item in self.bucket.objects.filter(Prefix=self._new_pfx):
-            yield self._get_journal(_JournalID.from_key(item.key))
-
-    def _generate_presigned_url(self, journal_id):
-        key = f"{self._blobs_pfx}/{journal_id.blob_id}"
-        return self.s3.meta.client.generate_presigned_url(ClientMethod="get_object",
-                                                          Params=dict(Bucket=self.bucket.name, Key=key))
-
-    def _delete_journals(self, journal_ids):
-        with S3Deleter(self.bucket) as s3d:
-            for journal_id in journal_ids:
-                s3d.delete(f"{self._journal_pfx}/{journal_id}")
-                s3d.delete(f"{self._blobs_pfx}/{journal_id.blob_id}")
-                s3d.delete(f"{self._new_pfx}/{journal_id}")
-
-    def _journal_ids(self, from_date=None, to_date=None):
-        # TODO: heuristic to find from_date in bucket listing -xbrianh
+    def list_journals(self, from_date: datetime=None, to_date: datetime=None) -> typing.Iterator[JournalID]:
         search_range = DateRange(from_date, to_date)
-        for item in self.bucket.objects.filter(Prefix=self._journal_pfx):
-            journal_id = _JournalID.from_key(item.key)
+        for journal_id in self._Journal.list():
             journal_range = DateRange(journal_id.start_date, journal_id.end_date)
             if journal_range in search_range:
                 yield journal_id
             elif journal_id.start_date in search_range.future:
                 break
 
+    def list_manifests(self,
+                       from_date: datetime=None,
+                       to_date: datetime=None) -> typing.Iterator[typing.Mapping[str, typing.Any]]:
+        for journal_id in self.list_journals(from_date, to_date):
+            manifest = self._Journal.from_id(journal_id).manifest()
+            manifest['data_url'] = self._generate_presigned_url(journal_id)
+            yield manifest
+
     def _destroy(self):
         with S3Deleter(self.bucket) as s3d:
             for item in self.bucket.objects.filter(Prefix=f"{self.root_prefix}/"):
                 s3d.delete(item.key)
 
-def replay_with_urls(url_info, from_date=None, to_date=None):
+def replay_with_manifest(manifest: dict, from_date: datetime=None, to_date: datetime=None) -> typing.Iterator[Event]:
     search_range = DateRange(from_date, to_date)
-    for urls in url_info:
-        manifest = urls['manifest']
-        for event_info in manifest['events']:
-            if datetime_from_timestamp(event_info['timestamp']) in search_range:
-                break
-        byte_range = f"bytes={event_info['offset']}-{manifest['size']-1}"
-        resp = requests.get(urls['events'], headers=dict(Range=byte_range), stream=True)
-        resp.raise_for_status()
-        for item in manifest['events']:
-            event_date = datetime_from_timestamp(item['timestamp'])
-            if event_date in search_range:
-                yield Event(item['event_id'], event_date, resp.raw.read(item['size']))
-            elif event_date in search_range.future:
-                break
-
-class FlashFloodException(Exception):
-    pass
-
-class FlashFloodJournalingError(FlashFloodException):
-    pass
-
-class FlashFloodEventNotFound(FlashFloodException):
-    pass
+    for event_info in manifest['events']:
+        if datetime_from_timestamp(event_info['timestamp']) in search_range:
+            break
+    byte_range = f"bytes={event_info['offset']}-{manifest['size']-1}"
+    resp = requests.get(manifest['data_url'], headers=dict(Range=byte_range), stream=True)
+    resp.raise_for_status()
+    for item in manifest['events']:
+        event_date = datetime_from_timestamp(item['timestamp'])
+        if event_date in search_range:
+            yield Event(item['event_id'], event_date, resp.raw.read(item['size']))
+        elif event_date in search_range.future:
+            break

--- a/flashflood/__init__.py
+++ b/flashflood/__init__.py
@@ -197,28 +197,28 @@ class FlashFlood:
             elif journal_id.start_date in search_range.future:
                 break
 
-    def list_manifests(self,
-                       from_date: datetime=None,
-                       to_date: datetime=None) -> typing.Iterator[typing.Mapping[str, typing.Any]]:
+    def list_event_streams(self,
+                           from_date: datetime=None,
+                           to_date: datetime=None) -> typing.Iterator[typing.Mapping[str, typing.Any]]:
         for journal_id in self.list_journals(from_date, to_date):
-            manifest = self._Journal.from_id(journal_id).manifest()
-            manifest['data_url'] = self._generate_presigned_url(journal_id)
-            yield manifest
+            event_stream = self._Journal.from_id(journal_id).manifest()
+            event_stream['stream_url'] = self._generate_presigned_url(journal_id)
+            yield event_stream
 
     def _destroy(self):
         with S3Deleter(self.bucket) as s3d:
             for item in self.bucket.objects.filter(Prefix=f"{self.root_prefix}/"):
                 s3d.delete(item.key)
 
-def replay_with_manifest(manifest: dict, from_date: datetime=None, to_date: datetime=None) -> typing.Iterator[Event]:
+def replay_event_stream(event_stream: dict, from_date: datetime=None, to_date: datetime=None) -> typing.Iterator[Event]:
     search_range = DateRange(from_date, to_date)
-    for event_info in manifest['events']:
+    for event_info in event_stream['events']:
         if datetime_from_timestamp(event_info['timestamp']) in search_range:
             break
-    byte_range = f"bytes={event_info['offset']}-{manifest['size']-1}"
-    resp = requests.get(manifest['data_url'], headers=dict(Range=byte_range), stream=True)
+    byte_range = f"bytes={event_info['offset']}-{event_stream['size']-1}"
+    resp = requests.get(event_stream['stream_url'], headers=dict(Range=byte_range), stream=True)
     resp.raise_for_status()
-    for item in manifest['events']:
+    for item in event_stream['events']:
         event_date = datetime_from_timestamp(item['timestamp'])
         if event_date in search_range:
             yield Event(item['event_id'], event_date, resp.raw.read(item['size']))

--- a/flashflood/__init__.py
+++ b/flashflood/__init__.py
@@ -7,11 +7,11 @@ from uuid import uuid4
 from concurrent.futures import ThreadPoolExecutor, as_completed
 
 from flashflood.util import datetime_to_timestamp, datetime_from_timestamp, DateRange, S3Deleter
-from flashflood.objects import BaseJournal, BaseJournalUpdate
+from flashflood.objects import Event, BaseJournal, BaseJournalUpdate
 from flashflood.identifiers import JournalID, TOMBSTONE_SUFFIX
-from flashflood.event import Event
 from flashflood.key_index import BaseKeyIndex
-from flashflood.exceptions import FlashFloodEventNotFound, FlashFloodJournalingError
+from flashflood.exceptions import (FlashFloodException, FlashFloodEventNotFound, FlashFloodEventExistsError,
+                                   FlashFloodJournalingError)
 
 
 # TODO:
@@ -60,17 +60,13 @@ class FlashFlood:
         self._KeyIndex = _KeyIndex
 
     def put(self, data, event_id: str=None, date: datetime=None) -> Event:
-        """
-        Write new event. If event exists, write update marker for event.
-        """
         date = date or datetime.utcnow()
         timestamp = datetime_to_timestamp(date)
         event_id = event_id or str(uuid4())
-        assert JournalID.DELIMITER not in event_id
+        if JournalID.DELIMITER in event_id:
+            raise FlashFloodException(f"'{JournalID.DELIMITER}' not allowed in event_id")
         if self.event_exists(event_id):
-            journal_id = self._journal_for_event(event_id)
-            self._JournalUpdate.upload_update(journal_id, event_id, data)
-            return Event(event_id, None, data)
+            raise FlashFloodEventExistsError(f"Event {event_id} already exists")
         else:
             events = [dict(event_id=event_id, timestamp=timestamp, offset=0, size=len(data))]
             journal = self._Journal(events, data=data, version="new")
@@ -83,7 +79,14 @@ class FlashFlood:
         journal_id = journal.id_
         self._KeyIndex.put_batch({e['event_id']: journal_id for e in journal.events})
 
-    def delete(self, event_id: str):
+    def update_event(self, event_id: str, new_data: bytes):
+        if self.event_exists(event_id):
+            journal_id = self._journal_for_event(event_id)
+            self._JournalUpdate.upload_update(journal_id, event_id, new_data)
+        else:
+            raise FlashFloodEventNotFound(f"Event {event_id} not found")
+
+    def delete_event(self, event_id: str):
         """
         Write delete marker for event.
         """

--- a/flashflood/event.py
+++ b/flashflood/event.py
@@ -1,3 +1,0 @@
-from collections import namedtuple
-
-Event = namedtuple("Event", "event_id date data")

--- a/flashflood/exceptions.py
+++ b/flashflood/exceptions.py
@@ -4,6 +4,9 @@ class FlashFloodException(Exception):
 class FlashFloodJournalingError(FlashFloodException):
     pass
 
+class FlashFloodEventExistsError(FlashFloodException):
+    pass
+
 class FlashFloodEventNotFound(FlashFloodException):
     pass
 

--- a/flashflood/objects.py
+++ b/flashflood/objects.py
@@ -1,6 +1,7 @@
 import io
 import json
 import typing
+from datetime import datetime
 from uuid import uuid4
 from string import hexdigits
 from collections import defaultdict, OrderedDict
@@ -8,9 +9,14 @@ from collections import defaultdict, OrderedDict
 from botocore.exceptions import ClientError
 
 from flashflood.util import datetime_from_timestamp, timestamp_now, S3Deleter
-from flashflood.event import Event
 from flashflood.exceptions import FlashFloodException, FlashFloodEventNotFound, FlashFloodJournalUploadError
 from flashflood.identifiers import JournalID, JournalUpdateID, JournalUpdateAction, TOMBSTONE_SUFFIX
+
+
+class Event(typing.NamedTuple):
+    event_id: str
+    date: datetime
+    data: bytes
 
 
 class BaseJournalUpdate:

--- a/requirements-dev.txt
+++ b/requirements-dev.txt
@@ -2,4 +2,5 @@ flake8
 coverage
 httpie
 twine
+mypy
 -r requirements.txt

--- a/tests/test_flashflood.py
+++ b/tests/test_flashflood.py
@@ -4,6 +4,7 @@ import sys
 from uuid import uuid4
 import unittest
 import boto3
+import json
 from concurrent.futures import ThreadPoolExecutor, as_completed
 from random import randint
 
@@ -11,25 +12,25 @@ pkg_root = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))  # noq
 sys.path.insert(0, pkg_root)  # noqa
 
 import flashflood
-from flashflood.util import datetime_from_timestamp
-from tests import infra
+from flashflood.util import datetime_from_timestamp, delete_keys
+from flashflood.exceptions import FlashFloodEventNotFound
+from tests import infra, random_date
 
 
 class TestFlashFlood(unittest.TestCase):
     @classmethod
     def setUpClass(cls):
-        cls.root_pfx = f"flashflood_test_{uuid4()}"
         cls.s3 = boto3.resource("s3")
         cls.bucket = cls.s3.Bucket(infra.get_env("S3_BUCKET"))
-        cls.flashflood = flashflood.FlashFlood(cls.s3, cls.bucket.name, cls.root_pfx)
 
     @classmethod
     def tearDownClass(cls):
-        with ThreadPoolExecutor(max_workers=10) as e:
-            futures = [e.submit(item.delete)
-                       for item in cls.bucket.objects.filter(Prefix="flashflood_test_")]
-            for f in as_completed(futures):
-                f.result()
+        keys = [item.key for item in cls.bucket.objects.filter(Prefix=f"flashflood_test_")]
+        delete_keys(cls.bucket, keys)
+
+    def setUp(self):
+        self.root_pfx = f"flashflood_test_{uuid4()}"
+        self.flashflood = flashflood.FlashFlood(self.s3, self.bucket.name, self.root_pfx)
 
     def tearDown(self):
         self.flashflood._destroy()
@@ -54,7 +55,65 @@ class TestFlashFlood(unittest.TestCase):
             with self.assertRaises(flashflood.FlashFloodEventNotFound):
                 self.flashflood.get_event("no_such_event")
 
-    def test_updates_and_deletes(self):
+    def test_updates_and_deletes_2(self):
+        number_of_events = 2
+        with self.subTest("Test update event"):
+            events = self.generate_events(number_of_events, journal=True)
+            event_id = set(events.keys()).pop()
+            new_event_data = self._random_data()
+            self.flashflood.put(new_event_data, event_id)
+            self._find_event_test(event_id, events[event_id].data)
+            self.flashflood.update()
+            self._find_event_test(event_id, new_event_data)
+        self.assertEqual(0, len([j for j in self.flashflood._new_journals()]))
+        with self.subTest("Test update new event"):
+            events = self.generate_events(number_of_events, journal=False)
+            event_id = set(events.keys()).pop()
+            new_event_data = self._random_data()
+            self.flashflood.put(new_event_data, event_id)
+            self._find_event_test(event_id, events[event_id].data)
+            self.flashflood.journal(minimum_number_of_events=number_of_events)
+            self._find_event_test(event_id, new_event_data)
+        self.assertEqual(0, len([j for j in self.flashflood._new_journals()]))
+        with self.subTest("Test delete event"):
+            events = self.generate_events(number_of_events, journal=True)
+            event_id = set(events.keys()).pop()
+            self.flashflood.delete(event_id)
+            self._find_event_test(event_id, should_find_event_in_replay=True, should_find_event_in_lookup=False)
+            self.flashflood.update()
+            self._find_event_test(event_id, should_find_event_in_replay=False, should_find_event_in_lookup=False)
+        self.assertEqual(0, len([j for j in self.flashflood._new_journals()]))
+        with self.subTest("Test delete new event"):
+            events = self.generate_events(number_of_events, journal=False)
+            event_id = set(events.keys()).pop()
+            self.flashflood.delete(event_id)
+            self._find_event_test(event_id, should_find_event_in_replay=True, should_find_event_in_lookup=False)
+            self.flashflood.journal(minimum_number_of_events=number_of_events)
+            self._find_event_test(event_id, should_find_event_in_replay=False, should_find_event_in_lookup=False)
+
+    def _find_event_test(self,
+                         event_id: str,
+                         expected_data: bytes=None,
+                         should_find_event_in_replay: bool=True,
+                         should_find_event_in_lookup: bool=True):
+        event_found = False
+        for event in self.flashflood.replay():
+            if event.event_id == event_id:
+                event_found = True
+                break
+        self.assertTrue(event_found == should_find_event_in_replay)
+        if expected_data is not None:
+            self.assertEqual(event.data, expected_data)
+
+        if should_find_event_in_lookup:
+            event = self.flashflood.get_event(event_id)
+            if expected_data is not None:
+                self.assertEqual(event.data, expected_data)
+        else:
+            with self.assertRaises(FlashFloodEventNotFound):
+                self.flashflood.get_event(event_id)
+
+    def test_updates_and_deletes_1(self):
         events = self.generate_events(15, journal=False)
         with self.subTest("Update event before journaling"):
             self._test_update(events)
@@ -72,9 +131,9 @@ class TestFlashFlood(unittest.TestCase):
 
     def _test_update(self, events, number_of_updates=2, number_of_events_to_journal=0):
         new_event_data = dict()
-        for _ in range(2):
+        for _ in range(number_of_updates):
             event_id = [event_id for event_id in events][randint(0, len(events) - 1)]
-            new_event_data[event_id] = os.urandom(5)
+            new_event_data[event_id] = self._random_data()
             self.flashflood.put(new_event_data[event_id], event_id)
         if 0 < number_of_events_to_journal:
             self.flashflood.journal(minimum_number_of_events=number_of_events_to_journal // 2)
@@ -89,7 +148,7 @@ class TestFlashFlood(unittest.TestCase):
 
     def _test_delete(self, events, number_of_deletes=2, number_of_events_to_journal=0):
         deleted_events = list()
-        for _ in range(2):
+        for _ in range(number_of_deletes):
             event_id = [event_id for event_id in events][randint(0, len(events) - 1)]
             deleted_events.append(event_id)
             del events[event_id]
@@ -100,9 +159,6 @@ class TestFlashFlood(unittest.TestCase):
             self.flashflood.journal(minimum_number_of_events=number_of_events_to_journal // 2)
         else:
             self.flashflood.update()
-        for event_id in deleted_events:
-            with self.assertRaises(flashflood.FlashFloodEventNotFound):
-                self.flashflood.get_event(event_id)
 
     def test_events(self):
         events = dict()
@@ -133,15 +189,13 @@ class TestFlashFlood(unittest.TestCase):
             from_date = ordered_dates[0]
             to_date = ordered_dates[-2]
             retrieved_events = list()
-            while True:
-                replay_urls = self.flashflood.replay_urls(from_date, to_date, maximum_number_of_results=1)
-                if not replay_urls:
-                    break
-                new_retrieved_events = [event for event in flashflood.replay_with_urls(replay_urls, from_date, to_date)]
-                for event in new_retrieved_events:
+            for manifest in self.flashflood.list_manifests(from_date, to_date):
+                new_retrieved_events = list()
+                for event in flashflood.replay_with_manifest(manifest, from_date, to_date):
                     self.assertGreater(event.date, from_date)
+                    new_retrieved_events.append(event)
                 retrieved_events.extend(new_retrieved_events)
-                from_date = datetime_from_timestamp(replay_urls[-1]['manifest']['to_date'])
+                from_date = datetime_from_timestamp(manifest['to_date'])
                 if not (from_date < to_date):
                     break
             self.assertEqual(len(dates) - 2, len(retrieved_events))
@@ -153,22 +207,23 @@ class TestFlashFlood(unittest.TestCase):
         events = self.generate_events(10)
         events = sorted([e for e in events.values()], key=lambda e: e.date)
         from_date = events[3].date
-        replay_urls = self.flashflood.replay_urls(from_date)
-        retrieved_events = [event for event in flashflood.replay_with_urls(replay_urls, from_date)]
+        retrieved_events = [event
+                            for manifest in self.flashflood.list_manifests(from_date)
+                            for event in flashflood.replay_with_manifest(manifest, from_date)]
         for event in events[:4]:
             self.assertNotIn(event, retrieved_events)
         for event in events[4:]:
             self.assertIn(event, retrieved_events)
 
     def test_journal(self):
-        self.generate_events(1, journal=False, event_size=1)
+        self.generate_events(1, journal=False)
         with self.subTest("raise FlashFloodJournalingError when attempting to journal more new events than available"):
             with self.assertRaises(flashflood.FlashFloodJournalingError):
                 self.flashflood.journal(minimum_number_of_events=2)
         with self.subTest("raise FlashFloodJournalingError when new event data doesn't meet size threshold"):
             with self.assertRaises(flashflood.FlashFloodJournalingError):
                 self.flashflood.journal(minimum_size=10)
-        self.generate_events(4, journal=False, event_size=1)
+        self.generate_events(4, journal=False)
         with self.subTest("Should succeed when minimum number and size thresholds are met"):
             self.flashflood.journal(minimum_number_of_events=5, minimum_size=5)
 
@@ -176,22 +231,22 @@ class TestFlashFlood(unittest.TestCase):
         events = dict()
         events.update(self.generate_events())
         events.update(self.generate_events())
-        replay_urls = self.flashflood.replay_urls(maximum_number_of_results=2)
         retrieved_events = {event.event_id: event
-                            for event in flashflood.replay_with_urls(replay_urls)}
+                            for manifest in self.flashflood.list_manifests()
+                            for event in flashflood.replay_with_manifest(manifest)}
         for event_id in events:
             self.assertEqual(events[event_id].data, retrieved_events[event_id].data)
 
     def test_get_new_journals(self):
         number_of_events = 3
         self.generate_events(number_of_events, journal=False)
-        new_journals = [c for c in self.flashflood._new_journals()]
+        new_journals = [j for j in self.flashflood._new_journals()]
         self.assertEqual(number_of_events, len(new_journals))
 
-    def generate_events(self, number_of_events=7, journal=True, event_size=3):
+    def generate_events(self, number_of_events=7, journal=True):
         def _put():
             event_id = str(uuid4()) + ".asdj__argh"
-            return self.flashflood.put(os.urandom(event_size), event_id, self._random_timestamp())
+            return self.flashflood.put(self._random_data(), event_id, random_date())
 
         with ThreadPoolExecutor(max_workers=10) as e:
             futures = [e.submit(_put) for _ in range(number_of_events)]
@@ -200,16 +255,8 @@ class TestFlashFlood(unittest.TestCase):
             self.flashflood.journal(minimum_number_of_events=number_of_events)
         return events
 
-    def _random_timestamp(self):
-        year = "%04i" % randint(1000, 2019)
-        month = "%02i" % randint(1, 12)
-        day = "%02i" % randint(1, 28)
-        hours = "%02i" % randint(0, 23)
-        minutes = "%02i" % randint(0, 59)
-        seconds = "%02i" % randint(0, 59)
-        fractions = "%06i" % randint(1, 999999)
-        timestamp = f"{year}-{month}-{day}T{hours}{minutes}{seconds}.{fractions}Z"
-        return datetime_from_timestamp(timestamp)
+    def _random_data(self, lower_size_limit=1, upper_size_limit=10) -> bytes:
+        return os.urandom(randint(lower_size_limit, upper_size_limit))
 
 if __name__ == '__main__':
     unittest.main()

--- a/tests/test_flashflood.py
+++ b/tests/test_flashflood.py
@@ -13,7 +13,7 @@ sys.path.insert(0, pkg_root)  # noqa
 
 import flashflood
 from flashflood.util import datetime_from_timestamp, delete_keys
-from flashflood.exceptions import FlashFloodEventNotFound
+from flashflood.exceptions import FlashFloodException, FlashFloodEventNotFound, FlashFloodEventExistsError
 from tests import infra, random_date
 
 
@@ -34,6 +34,18 @@ class TestFlashFlood(unittest.TestCase):
 
     def tearDown(self):
         self.flashflood._destroy()
+
+    def test_put_and_update_event_exceptions(self):
+        events = self.generate_events(1, journal=False)
+        with self.subTest("Should not be able to put event_id with illegal chars"):
+            with self.assertRaises(FlashFloodException):
+                self.flashflood.put(b"", "asldkfj--lasjf")
+        with self.subTest("Should not be able to put an existing event"):
+            with self.assertRaises(FlashFloodEventExistsError):
+                self.flashflood.put(b"", next(iter(events.values())).event_id)
+        with self.subTest("Should not be able to update a non-existent event"):
+            with self.assertRaises(FlashFloodEventNotFound):
+                self.flashflood.update_event(str(uuid4()), b"")
 
     def test_get_event(self):
         events = self.generate_events(10, journal=False)
@@ -61,7 +73,7 @@ class TestFlashFlood(unittest.TestCase):
             events = self.generate_events(number_of_events, journal=True)
             event_id = set(events.keys()).pop()
             new_event_data = self._random_data()
-            self.flashflood.put(new_event_data, event_id)
+            self.flashflood.update_event(event_id, new_event_data)
             self._find_event_test(event_id, events[event_id].data)
             self.flashflood.update()
             self._find_event_test(event_id, new_event_data)
@@ -70,7 +82,7 @@ class TestFlashFlood(unittest.TestCase):
             events = self.generate_events(number_of_events, journal=False)
             event_id = set(events.keys()).pop()
             new_event_data = self._random_data()
-            self.flashflood.put(new_event_data, event_id)
+            self.flashflood.update_event(event_id, new_event_data)
             self._find_event_test(event_id, events[event_id].data)
             self.flashflood.journal(minimum_number_of_events=number_of_events)
             self._find_event_test(event_id, new_event_data)
@@ -78,7 +90,7 @@ class TestFlashFlood(unittest.TestCase):
         with self.subTest("Test delete event"):
             events = self.generate_events(number_of_events, journal=True)
             event_id = set(events.keys()).pop()
-            self.flashflood.delete(event_id)
+            self.flashflood.delete_event(event_id)
             self._find_event_test(event_id, should_find_event_in_replay=True, should_find_event_in_lookup=False)
             self.flashflood.update()
             self._find_event_test(event_id, should_find_event_in_replay=False, should_find_event_in_lookup=False)
@@ -86,7 +98,7 @@ class TestFlashFlood(unittest.TestCase):
         with self.subTest("Test delete new event"):
             events = self.generate_events(number_of_events, journal=False)
             event_id = set(events.keys()).pop()
-            self.flashflood.delete(event_id)
+            self.flashflood.delete_event(event_id)
             self._find_event_test(event_id, should_find_event_in_replay=True, should_find_event_in_lookup=False)
             self.flashflood.journal(minimum_number_of_events=number_of_events)
             self._find_event_test(event_id, should_find_event_in_replay=False, should_find_event_in_lookup=False)
@@ -134,7 +146,7 @@ class TestFlashFlood(unittest.TestCase):
         for _ in range(number_of_updates):
             event_id = [event_id for event_id in events][randint(0, len(events) - 1)]
             new_event_data[event_id] = self._random_data()
-            self.flashflood.put(new_event_data[event_id], event_id)
+            self.flashflood.update_event(event_id, new_event_data[event_id])
         if 0 < number_of_events_to_journal:
             self.flashflood.journal(minimum_number_of_events=number_of_events_to_journal // 2)
             self.flashflood.update()
@@ -152,7 +164,7 @@ class TestFlashFlood(unittest.TestCase):
             event_id = [event_id for event_id in events][randint(0, len(events) - 1)]
             deleted_events.append(event_id)
             del events[event_id]
-            self.flashflood.delete(event_id)
+            self.flashflood.delete_event(event_id)
         if 0 < number_of_events_to_journal:
             self.flashflood.journal(minimum_number_of_events=number_of_events_to_journal // 2)
             self.flashflood.update()

--- a/tests/test_flashflood.py
+++ b/tests/test_flashflood.py
@@ -189,9 +189,9 @@ class TestFlashFlood(unittest.TestCase):
             from_date = ordered_dates[0]
             to_date = ordered_dates[-2]
             retrieved_events = list()
-            for manifest in self.flashflood.list_manifests(from_date, to_date):
+            for manifest in self.flashflood.list_event_streams(from_date, to_date):
                 new_retrieved_events = list()
-                for event in flashflood.replay_with_manifest(manifest, from_date, to_date):
+                for event in flashflood.replay_event_stream(manifest, from_date, to_date):
                     self.assertGreater(event.date, from_date)
                     new_retrieved_events.append(event)
                 retrieved_events.extend(new_retrieved_events)
@@ -208,8 +208,8 @@ class TestFlashFlood(unittest.TestCase):
         events = sorted([e for e in events.values()], key=lambda e: e.date)
         from_date = events[3].date
         retrieved_events = [event
-                            for manifest in self.flashflood.list_manifests(from_date)
-                            for event in flashflood.replay_with_manifest(manifest, from_date)]
+                            for manifest in self.flashflood.list_event_streams(from_date)
+                            for event in flashflood.replay_event_stream(manifest, from_date)]
         for event in events[:4]:
             self.assertNotIn(event, retrieved_events)
         for event in events[4:]:
@@ -227,13 +227,13 @@ class TestFlashFlood(unittest.TestCase):
         with self.subTest("Should succeed when minimum number and size thresholds are met"):
             self.flashflood.journal(minimum_number_of_events=5, minimum_size=5)
 
-    def test_urls(self):
+    def test_event_streams(self):
         events = dict()
         events.update(self.generate_events())
         events.update(self.generate_events())
         retrieved_events = {event.event_id: event
-                            for manifest in self.flashflood.list_manifests()
-                            for event in flashflood.replay_with_manifest(manifest)}
+                            for event_stream in self.flashflood.list_event_streams()
+                            for event in flashflood.replay_event_stream(event_stream)}
         for event_id in events:
             self.assertEqual(events[event_id].data, retrieved_events[event_id].data)
 

